### PR TITLE
Log warning if applicationset contains unrecognized generators

### PR DESCRIPTION
**Pull request by jopit**

### CODE REVIEW

File: pkg/controllers/applicationset_controller.go [hunk 1]
```diff
@@ -67,6 +70,23 @@ func (r *ApplicationSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	// Log a warning if there are unrecognized generators
```
Comment (jgwest) [inline pkg/controllers/applicationset_controller.go:73]: Nitpick: I think this block can be moved into its own function (eg `checkInvalidGenerators`), which can then be called from here.

File: pkg/controllers/applicationset_controller.go [hunk 2]
```diff
@@ -67,6 +70,23 @@ func (r *ApplicationSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
...
+	hasInvalidGenerators, invalidGenerators := invalidateGenerators(&applicationSetInfo)
+	if len(invalidGenerators) > 0 {
+		gnames := make([]string, 0, len(invalidGenerators))
```
Comment (jgwest) [inline pkg/controllers/applicationset_controller.go:76]: Nitpick: this is perfectly valid but `gnames := []string{}` is a bit cleaner

File: pkg/controllers/applicationset_controller.go [hunk 3]
```diff
@@ -100,6 +120,68 @@ func (r *ApplicationSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
...
+	generator, ok := generators[index].(map[string]interface{})
+	if !ok {
+		log.Warnf("coundn't get generator from kubectl.kubernetes.io/last-applied-configuration annotation")
+		return
+	}
+
+	var name string
+	for key := range generator {
+		name = key
+		names[name] = true
```
Comment (jgwest) [inline pkg/controllers/applicationset_controller.go:180]: Nitpick: think this can just be `names[key] = true`

**Review by jgwest — CHANGES_REQUESTED** [review]
Thanks @jopit, code LGTM, just some nitpicks I've mentioned below... It would also be good to add a unit test for this case, for example testing the `invalidateGenerators` function. (Once https://github.com/argoproj-labs/applicationset/pull/64/files merges then tests runs will be included in the PR build with a `make test`, but in the mean time we have to do `go test ./...`)

**Comment by jopit** [comment]
@jgwest thanks for the review.  I've made changes based on your feedback.

File: pkg/controllers/applicationset_controller.go (hunk 2 above)
Comment (jopit) [reply pkg/controllers/applicationset_controller.go]: Done

File: pkg/controllers/applicationset_controller.go (hunk 3 above)
Comment (jopit) [reply pkg/controllers/applicationset_controller.go]: Done

File: pkg/controllers/applicationset_controller.go (hunk 1 above)
Comment (jopit) [reply pkg/controllers/applicationset_controller.go:73]: Done

**Review by jgwest — APPROVED** [review]
Thanks @jopit, LGTM and approved!

**Review by xianlubird — APPROVED** [review]

**Pull request merged**